File: solutions/elixir/roman-numerals/2/lib/roman_numerals.ex
defmodule RomanNumerals do
  defp get_roman_numerals(base, five, ten, times) do
    case times do
      9 -> base <> ten
      times when times >= 5 -> five <> String.duplicate(base, times - 5)
      4 -> base <> five
      _ -> String.duplicate(base, times)
    end
  end

  @doc """
  Convert the number to a roman number.
  """
  @spec numeral(pos_integer) :: String.t()
  def numeral(number) do
    case number do
      number when number > 1000 -> get_roman_numerals("M", "", "", div(number, 1000)) <> numeral(rem(number, 1000))
      number when number > 100 -> get_roman_numerals("C", "D", "M", div(number, 100)) <> numeral(rem(number, 100))
      number when number > 10 -> get_roman_numerals("X", "L", "C", div(number, 10)) <> numeral(rem(number, 10))
      _ -> get_roman_numerals("I", "V", "X", number)
    end
  end
end

File: solutions/python/roman-numerals/2/roman_numerals.py
def roman(number):
    DICT = {
        1000: {
            "one": "M",
            "five": "",
            "ten": "",
        },
        100: {
            "one": "C",
            "five": "D",
            "ten": "M",
        },
        10: {
            "one": "X",
            "five": "L",
            "ten": "C",
        },
        1: {
            "one": "I",
            "five": "V",
            "ten": "X"
        },
    }

    result = ""
    for divisor in DICT:
        if number >= divisor:
            times = number // divisor
            if times == 9:
                result += DICT[divisor]["one"] + DICT[divisor]["ten"]
            elif times >= 5:
                result += DICT[divisor]["five"] + DICT[divisor]["one"] * (times - 5)
            elif times == 4:
                result += DICT[divisor]["one"] + DICT[divisor]["five"]
            else:
                result += DICT[divisor]["one"] * times
            number -= times * divisor

    return result
